### PR TITLE
[WIP] dependency prioritization 

### PR DESCRIPTION
**Pull request by xu-cheng**
TODO list:

* [ ] perfromance evaluation 
* [ ] add test case

### CODE REVIEW

**Comment by xu-cheng** [comment]
cc @CNA-Bld

**Comment by xu-cheng** [comment]
I just realize this approach has tons of problems. Investigating better solution.

File: Library/Homebrew/dependency.rb [hunk 1]
```diff
@@ -2,9 +2,22 @@
 
 # A dependency on another Homebrew formula.
 class Dependency
+  class DependencyMergeConflict < RuntimeError
+    def initialize(formulae)
+      super <<-EOS.undent
+        Cannot merge dependencies, multiple conflicted formluae are requested.
```
Comment (mikemcquaid) [inline Library/Homebrew/dependency.rb]: `formluae` -> `formulae`

Comment (mikemcquaid) [inline Library/Homebrew/dependency.rb]: Also probably `conflicted` -> `conflicting`

**Comment by mikemcquaid** [comment]
Approach makes sense to me so far :+1:

**Comment by davydden** [comment]
@xu-cheng do you think you will have a chance to look into this PR in some near future? Would be a nice feature to have...

**Comment by xu-cheng** [comment]
@davydden this is on my todo list. Current PR has performance issue. And I'm working another PR to address this issue. So it will take a period of time before this finally landing.

**Comment by davydden** [comment]
@xu-cheng thanks for the prompt reply and the status update.

**Comment by UniqMartin** [comment]
Could you please summarize what the current behavior is (i.e., prior to this PR) and how and why this PR changes this? (I think this should go into the first post.) I would like to understand more of the dependency resolution logic, but all of this stuff is not exactly self-documenting. Thanks!

**Comment by xu-cheng** [comment]
## Current Homebrew's Behaviors 

* `depends_on` only invokes `Formulary.factory`. Therefore, for `depends_on "foo"`, it will look for core formula first, then in taps. If multiple tap formulae are found, `TapFormulaAmbiguityError` will be raised.
* `brew tap-pin` is only used to prioritize certain taps over core formula in `ARGV.formulae`. i.e. only help to offer shortcut for user's input. **Notice** this is the desired behavior. Because we believe formula resolution order by `depends_on` should be decided by formula authors instead of users.

## Desired Homebrew's Behaviors 

We need to change `depends_on "foo"`  to use following prioritization order:
* Look for formula inside its own tap. i.e. `homebrew/homebrew/foo` for core formulae and `some/tap/foo` for tap formulae.
* Look for formula inside core. This is mainly for tap formulae.
* Look for formula anywhere. A warning message will be shown.

## What need to be done(What holds this PR)

Current implementation has poor performance because we lack of tap information. I'm still working on passing tap information during loading formula code. And I'm still not satisfied with them. I'm hoping to find a delicate way to get formula file path during loading.

**Comment by UniqMartin** [comment]
Thanks for this overview/status, @xu-cheng! That's massively helpful to me (and I'm sure to others, too).

So the reason the current system works quite well is that at least the official taps take care to use unique formula names and thus the search order usually doesn't matter. Will there still be a possibility for a tap formula like `some/tap/formula` to depend on a core formula if there's also one with the same name in the tap, e.g., by specifying `homebrew/homebrew/dependency` so that it is chosen before `some/tap/dependency` from the same tap?

**Comment by mikemcquaid** [comment]
Agreed with the desired behavkours :+1:

**Comment by MikeMcQuaid** [comment]
Can this be ported over to the Homebrew/brew repo when it's ready to 🚢? Thanks!

**Pull request closed without merging**